src/storage/src/hummock/local_version_manager.rs

```diff
@@ -511,35 +508,92 @@ impl LocalVersionManager {
+                .read()
+                .iter_shared_buffer()
+                .find(|(key, _)| key > &&self.max_sync_epoch.load(Relaxed))
+                .map(|(key, _)| key)
+                .cloned();
+            if min_unsynced_epoch.unwrap() == epoch {
+                break;
+            } else {
+                sync_epoch_notify_clone.notified().await;
             }
+        }
+        let (uncommitted_data, task_write_batch_size) = {
+            // We keep the lock on max_sync_epoch until the task is saved in the sync task vec.
+            let mut local_version_guard = self.local_version.write();
+            self.max_sync_epoch.store(epoch, Relaxed);
+            let (uncommitted_data, task_write_batch_size) = match local_version_guard
+                .get_mut_shared_buffer(epoch)
+                .unwrap()
+                .take_uncommitted_data()
+            {
+                Some(task) => task,
+                None => {
+                    tracing::trace!("sync epoch {} has no more task to do", epoch);
+                    return Ok((0, vec![]));
+                }
+            };
+            local_version_guard.add_sync_state(epoch, Syncing(uncommitted_data.clone()));
+            (uncommitted_data, task_write_batch_size)
         };
 
-        self.run_upload_task(order_index, epoch, task_payload)
-            .await?;
+        self.sync_epoch_notify.notify_waiters();
+
         tracing::trace!(
             "sync epoch {} finished. Task size {}",
             epoch,
             task_write_batch_size
         );
+        let ssts = self
+            .sync_shared_buffer_epoch(epoch, uncommitted_data)
+            .await?;
+        Ok((task_write_batch_size, ssts))
+    }
+
+    pub async fn sync_shared_buffer_epoch(
+        &self,
+        epoch: HummockEpoch,
+        uncommitted_data: KeyIndexedUncommittedData,
+    ) -> HummockResult<Vec<LocalSstableInfo>> {
+        let task_payload = to_order_sorted(&uncommitted_data);
+        let ssts = self.run_sync_upload_task(epoch, task_payload).await?;
         if let Some(conflict_detector) = self.write_conflict_detector.as_ref() {
             conflict_detector.archive_epoch(epoch);
         }
         self.buffer_tracker
             .send_event(SharedBufferEvent::EpochSynced(epoch));
-        Ok(task_write_batch_size)
+        Ok(ssts)
+    }
+
+    pub async fn run_sync_upload_task(
+        &self,
+        epoch: HummockEpoch,
+        task_payload: UploadTaskPayload,
+    ) -> HummockResult<Vec<LocalSstableInfo>> {
+        let task_result = self.shared_buffer_uploader.flush(epoch, task_payload).await;
+        let mut local_version_guard = self.local_version.write();
+        let ssts = task_result?;
+
+        let mut new_sst = BTreeMap::default();
+        for sst in ssts.clone() {
+            new_sst.insert(
+                get_sst_key_range(&sst.1).right.as_slice().to_vec(),
+                UncommittedData::Sst(sst),
+            );
+        }
```

Comment:
Do we need to order SSTs by end key in `SyncUncommittedData::Synced`? SSTs are stored in a Vec in `SyncUncommittedData::Synced` and the end key ordering seems not to be used now. Can we simplify the logic by only wrapping SST in a Vec<UncommittedData::Sst>?